### ISSUE TITLE
storage: TestRebalance_3To5Small timing out

### ISSUE THREAD

**Issue by cuongdo** [cockroachdb/cockroach#8761]
The test is taking longer than the allotted 3 hours. When we first started running the test, it took about an hour to run.

https://teamcity.cockroachdb.com/viewLog.html?buildId=13814&buildTypeId=Cockroach_AllocatorTest3to510g&tab=buildLog#_focus=3369

Generally, the test had a > 80% pass rate 7 weeks ago. Now, it rarely passes.

cc @cockroachdb/stability

**Comment by cuongdo** [comment]
This is probably related to #8738, but I'm not sure. There are different symptoms, so I filed a separate issue.

**Comment by tamird** [comment]
Looking at the test's history, it seems that today's timeout was the first since July 29.

This test passed as recently as August 16. Do you have any ideas regarding what changed since then?

https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Cockroach_AllocatorTest3to510g&tab=buildTypeHistoryList

**Comment by cuongdo** [comment]
The pass on 8/16 was probably a fluke. Even before, the time required for the test to pass was fairly variable. I hadn't seen it time out in a while. However, this combined with the `context deadline exceeded` errors makes me think that there's something generally making rebalancing much slower than it should be.

So much has changed in the last month (and even week) that I couldn't really guess with any confidence.

**Comment by cuongdo** [comment]
The latest run failed:

https://teamcity.cockroachdb.com/viewLog.html?buildId=18755&buildTypeId=Cockroach_AllocatorTest3to510g&tab=buildLog

I haven't had a chance to dig deeply, but from glancing at the admin UI during the test, it looked like the replica counts were almost the same across the nodes but thrashing before they fully converged.

It looks like #8959 has significantly changed the allocator, so it might make sense to run the test against `develop` to see things have improved.

**Comment by bdarnell** [comment]
I saw this thrashing in production too (#9031).

**Comment by cuongdo** [comment]
same thrashing exists on `develop`

**Comment by petermattis** [comment]
Pretty sure the thrashing is caused by the balancer heuristics, not the candidate selection rules (which is what Tristan's work affected).

**Comment by cuongdo** [comment]
Yes. Just wanted to confirm before starting my changes.
